### PR TITLE
Smarter autosigner

### PR DESCRIPTION
**Pull request by diranged**
Instead of being dead-set on the exact layout of the CSR, search the CSR
properly for the bits and then compare them. No need to ensure that the
CSR has exactly the right number of attributes... nor to assume the exact
order in which they will show up.

### CODE REVIEW

File: lib/etc/autosign.rb [hunk 1]
```diff
@@ -78,19 +79,14 @@ def initialize()
...
-        key = extension.value[0].value
-        val = extension.value[1].value
-
-        if key == PP_PRESHARED_KEY_VALUE || key == 'pp_preshared_key'
-          return val
-        end
-      end
+      @log.info("Searching for #{key}")
+      attr = attributes.find{|x| x.oid == key}
+      attr.value.first.value
```
Comment (nehalrp) [inline lib/etc/autosign.rb]: Unless you're at the very end of the method, you should explicitly say `return attr.value.first.value`.

File: lib/etc/autosign.rb [hunk 2]
```diff
@@ -78,19 +79,14 @@ def initialize()
   # * *Returns* :
   #   - The value of the attribute matching the pp_preshared_key name.
   #
-  def find_preshared_key(attributes)
+  def find_attribute(key, attributes)
```
Comment (nehalrp) [inline lib/etc/autosign.rb]: You'll need to update the rspecs:
https://github.com/Nextdoor/puppet_rightscale/blob/matt/better_autosigner/spec/unit/etc/autosigner_spec.rb

File: lib/etc/autosign.rb [hunk 3]
```diff
@@ -78,19 +79,14 @@ def initialize()
...
-        if key == PP_PRESHARED_KEY_VALUE || key == 'pp_preshared_key'
-          return val
-        end
-      end
+      @log.info("Searching for #{key}")
+      attr = attributes.find{|x| x.oid == key}
+      attr.value.first.value
     rescue Exception => e
-      raise("Invalid CSR: #{e}")
+      # Pass
```
Comment (nehalrp) [inline lib/etc/autosign.rb]: We can get rid of exception handling here entirely if we're looking for keys? We'd have to handle `find_attribute ` returning `nil` then.

**Comment by diranged** [comment]
Ok I've fixed up all the broken tests - and refactored a bit more. Going to merge this as is, knowing that I hope to kill the autosigner entirely soon, or at least refactor it completely into a bash-script that uses `rsc`.

**Pull request merged**